blockchain_publisher/src/main/java/org/cardanofoundation/lob/app/blockchain_publisher/service/WatchDogService.java

```diff
@@ -0,0 +1,137 @@
+    private final ReportEntityRepositoryGateway reportEntityRepositoryGateway;
+
+    @Value("${lob.blockchain_publisher.watchdog.rollback.grace.period.minutes:15}")
+    @Getter
+    @Setter
+    private int rollbackGracePeriodMinutes = 15;
+
+    @Value("${lob.blockchain_publisher.watchdog.rollbacks.enabled:true}")
+    @Getter
+    @Setter
+    private boolean rollbackSupportEnabled = true;
+
+    @PostConstruct
+    public void init() {
+        log.info("TransactionsWatchDogService configuration: rollbackGracePeriodMinutes={}, rollbacksEnabled={}", rollbackGracePeriodMinutes, rollbackSupportEnabled);
+
+        log.info("TransactionsWatchDogService started");
+    }
+
+    public void checkTxStatusesForOrganisations(int txStatusInspectionLimitPerOrgPullSize) {
+        organisationPublicApiIF.listAll().forEach(org -> {
+            log.info("Checking transaction statuses for organisation: {}", org.getName());
+            checkTransactionStatusesForOrganisation(org, txStatusInspectionLimitPerOrgPullSize);
+        });
+    }
+
+    private void checkTransactionStatusesForOrganisation(Organisation org, int txStatusInspectionLimitPerOrgPullSize) {
+        ChainTip chainTip = getChainTip();
+        if (!chainTip.isSynced()) {
+            log.info("Chain is not synced, skipping transaction status check for organisation: {}", org.getName());
+            return;
+        }
+
+        transactionEntityRepositoryGateway.findDispatchedTransactionsThatAreNotFinalizedYet(org.getId(), Limit.of(txStatusInspectionLimitPerOrgPullSize)).forEach(tx -> {
+            log.info("Checking transaction status for transaction: {}", tx.getId());
+            L1SubmissionData l1SubmissionData1 = tx.getL1SubmissionData().orElseThrow(() -> new RuntimeException("Failed to get L1 submission data"));
+            tx.setL1SubmissionData(Optional.of(updateL1SubmissionData(l1SubmissionData1, chainTip)));
+
+            transactionEntityRepositoryGateway.storeTransaction(tx);
+            log.info("Status updated for transaction: {}", tx.getId());
+        });
+
+        reportEntityRepositoryGateway.findDispatchedReportsThatAreNotFinalizedYet(org.getId(), Limit.of(txStatusInspectionLimitPerOrgPullSize)).forEach(report -> {
+            log.info("Checking transaction status for report: {}", report.getId());
+            L1SubmissionData l1SubmissionData = report.getL1SubmissionData().orElseThrow(() -> new RuntimeException("Failed to get L1 submission data"));
+            report.setL1SubmissionData(Optional.of(updateL1SubmissionData(l1SubmissionData, chainTip)));
+
+            reportEntityRepositoryGateway.storeReport(report);
+            log.info("Status updated for report: {}", report.getId());
+        });
+    }
+
+    private OnChainStatus getOnChainStatus(Optional<OnChainTxDetails> onChainTxDetails, Long txCreationSlot, ChainTip chainTip) {
+        if (onChainTxDetails.isPresent()) {
```

Comment:
if you have if else with return you don't need to have else because return is already returning